drivers/setup/blue/blue.c

```diff
@@ -1016,83 +1163,134 @@ ScrIoControl(
                 vidmem = DeviceExtension->VideoMemory;
                 offset = (Buf->dwCoord.X + Buf->dwCoord.Y * DeviceExtension->Columns) * 2 + 1;
 
-                nMaxLength = min(stk->Parameters.DeviceIoControl.OutputBufferLength,
+                nMaxLength = min(nMaxLength,
                                  (DeviceExtension->Rows - Buf->dwCoord.Y)
                                     * DeviceExtension->Columns - Buf->dwCoord.X);
 
                 for (dwCount = 0; dwCount < nMaxLength; dwCount++, pAttr++)
                 {
                     *((PCHAR)pAttr) = vidmem[offset + (dwCount * 2)];
                 }
-
-                Buf->dwTransfered = dwCount;
-            }
-            else
-            {
-                Buf->dwTransfered = 0;
+                Irp->IoStatus.Information = dwCount * sizeof(USHORT);
             }
 
-            Irp->IoStatus.Information = sizeof(OUTPUT_ATTRIBUTE);
             Status = STATUS_SUCCESS;
             break;
         }
 
         case IOCTL_CONSOLE_WRITE_OUTPUT_ATTRIBUTE:
         {
-            PCOORD pCoord = (PCOORD)MmGetSystemAddressForMdl(Irp->MdlAddress);
-            PCHAR pAttr = (PCHAR)(pCoord + 1);
+            PCOORD pCoord;
+            PUSHORT pAttr;
             PUCHAR vidmem;
             ULONG offset;
             ULONG dwCount;
             ULONG nMaxLength;
 
+            //
+            // NOTE: For whatever reason no OUTPUT_ATTRIBUTE structure
+            // is used for this IOCTL.
+            //
+
+            /* Validate output buffer */
+            if (stk->Parameters.DeviceIoControl.OutputBufferLength < sizeof(COORD))
+            {
+                Status = STATUS_INVALID_PARAMETER;
+                break;
+            }
+            ASSERT(Irp->MdlAddress);
+            pCoord = (PCOORD)MmGetSystemAddressForMdl(Irp->MdlAddress);
+            if (pCoord == NULL)
+            {
+                Status = STATUS_INSUFFICIENT_RESOURCES;
+                break;
+            }
+
+            nMaxLength = stk->Parameters.DeviceIoControl.OutputBufferLength - sizeof(COORD);
+            Irp->IoStatus.Information = 0;
+
             if ( pCoord->X < 0 || pCoord->X >= DeviceExtension->Columns ||
-                 pCoord->Y < 0 || pCoord->Y >= DeviceExtension->Rows )
+                 pCoord->Y < 0 || pCoord->Y >= DeviceExtension->Rows    ||
+                 nMaxLength == 0 )
             {
-                Irp->IoStatus.Information = 0;
                 Status = STATUS_SUCCESS;
                 break;
             }
 
+            pAttr = (PUSHORT)(pCoord + 1);
+
             if (DeviceExtension->Enabled && DeviceExtension->VideoMemory)
             {
                 vidmem = DeviceExtension->VideoMemory;
                 offset = (pCoord->X + pCoord->Y * DeviceExtension->Columns) * 2 + 1;
 
-                nMaxLength = min(stk->Parameters.DeviceIoControl.OutputBufferLength - sizeof(COORD),
+                nMaxLength = min(nMaxLength,
```

Comment:
Since you changed `pAttr` to be a `PUSHORT`, you're now increasing it by two bytes each loop iteration. So you can only do `nMaxLength / sizeof(USHORT)` iterations.